### ISSUE TITLE
[Bug]: `plot_surface` changes color after transposing the input matrices

### ISSUE THREAD

**Issue by axil** [matplotlib/matplotlib#23570]
### Bug summary

When I transpose X, Y, and Z, the color of the surface changes from light blue to dark blue (other colors darken as well).

### Code for reproduction

```python
import matplotlib.pyplot as plt
fig, ax = plt.subplots(1, 2, subplot_kw={"projection": "3d"}, figsize=[18,10])
x = np.arange(2);
y = np.arange(2);
X, Y = np.meshgrid(x, y)
ax[0].plot_surface(X, Y, np.ones((2,2)));
ax[1].plot_surface(X.T, Y.T, np.ones((2,2)));
```

### Actual outcome

![image](https://user-images.githubusercontent.com/170910/183258503-628a866e-392d-468c-a2ce-d8c8660b0cc4.png)

### Expected outcome

Surfaces of the same color

### Additional information

I tried several backends, several python and matplotlib versions, the effect is the same.
It is somehow connected with shading because when I write `shade=False` the colors are the same.
But as far as I understand shading should not depend on the order of the points.

### Operating system

Windows 10

### Matplotlib Version

3.5.2, 3.5.1

### Matplotlib Backend

notebook, qt

### Python version

3.7.7, 3.9

### Jupyter version

6.0.1

### Installation

pip

**Comment by oscargus** [comment]
The difference is in the sign of the normals generated here:
https://github.com/matplotlib/matplotlib/blob/bf9a451bf08107cecb7f1b732ebfaa42c5dda133/lib/mpl_toolkits/mplot3d/axes3d.py#L1469-L1470

```
In [22]: ax[0]._generate_normals(p.vertices)
Out[22]: array(0.00550837)

In [23]: ax[1]._generate_normals(p2.vertices)
Out[23]: array(-0.00550837)
```
where p and p2 are the paths of the rectangles:
```
In [24]: p.vertices
Out[24]: 
array([[-0.07215045, -0.00966633],
       [ 0.02025935, -0.03780446],
       [ 0.06981276,  0.00935314],
       [-0.01791343,  0.03342692],
       [-0.07215045, -0.00966633]])

In [25]: p2.vertices
Out[25]: 
array([[-0.07215045, -0.00966633],
       [-0.01791343,  0.03342692],
       [ 0.06981276,  0.00935314],
       [ 0.02025935, -0.03780446],
       [-0.07215045, -0.00966633]])
```
so they clearly are the same points, but in different order.

I guess it is a question of definition of the normal direction. On one hand one can put it to the absolute value as your example clearly is confusing, on the other hand there may be situations where this breaks existing code and where one would like to be able to spot the difference.

**Comment by axil** [comment]
I would accept it as expected behaviour if one side would be 'lit' with normal color and the other would be darker. But as a matter of fact both sides are of the same color (in both cases). Why would it be?

**Comment by oscargus** [comment]
A guess is that the magnitude of the normal is pretty small, so one would expect a pretty small difference in shade. Positioning a light source differently would change the magnitude of the normal and therefore the difference in shade.

**Comment by axil** [comment]
Ok. I agree that this is an artificial example. I provided is only for the sake making the example minimally reproducible. 
How about this?
![image](https://user-images.githubusercontent.com/170910/183287011-f978c782-023a-41fa-a409-3ede0ff4a07a.png)
I'm plotting a surface, I'm asking it to be green. It is green when X and Y are in the 'proper' order. When they are transposed, it is hardly green anymore.

This effect has gained little notice most probably because normally no one bothers to transpose the arguments.

**Comment by oscargus** [comment]
Yeah, I cannot really say that the 3D-rendering is very advanced. Nor do I have any idea how the `shade` argument is really supposed to work.

The practical question is really if it will be better or worse for the random user to apply abs to the computed normal. (Or if one should change the path generation so that the same path is always generated indendent of the input being transposed or not.)

**Comment by axil** [comment]
This definitely looks like a bug for me.

``` python
from math import pi
fig, ax = plt.subplots(1, 2, subplot_kw={"projection": "3d"}, figsize=[18,10])
x = np.linspace(-2*pi, 2*pi, 100)
y = np.linspace(-2*pi, 2*pi, 100)
X, Y = np.meshgrid(x, y)
ax[0].plot_surface(X, Y, np.sin(X)*np.sin(Y), color='red');
ax[1].plot_surface(X.T, Y.T, (np.sin(X)*np.sin(Y)).T, color='red');
```

![image](https://user-images.githubusercontent.com/170910/183297672-9fe3995b-263d-4285-b394-65992eadee9b.png)
In my opinion either the documentation should state clearly that the X and Y should be generated with `indexing='xy'` or the code should detect that the coordinates are given in the 'incorrect' order and compensate for it.

What is the default location of the light source? (-∞, -∞, 0) ?

**Comment by jklymak** [comment]
I think the normal the other way is just a simple way to make something look like it is on the inside of a shape.  But we don't have real ray tracing, so that simple way is not particularly physical.

**Comment by axil** [comment]
Setting `shading=False` as @oscargus suggested on SO is not a universal solution as it disables shading entirely:

![image](https://user-images.githubusercontent.com/170910/183299139-0e154471-e793-494e-8dd1-9c9643b70cf5.png)

@jklymak The thing I don't understand is why the surface has the same color from the upside and from the downside?

**Comment by jklymak** [comment]
Because if the surface is upside down, it is assumed to be the inside of a larger surface that has a portion that is right side up.  Think of a cube with one of the sides missing so you can see the bottom of the cube.  We can't shade that property with diffuse light so it seems we just shade it arbitrarily darker

**Comment by oscargus** [comment]
> is not a universal solution

Well, it is universal in the sense that transposition doesn't matter. ;-)

**Comment by oscargus** [comment]
A bit more on the actual source of the problem:
https://github.com/matplotlib/matplotlib/blob/cf38c3464ebdbc66eee97222445f9b861ac1a395/lib/mpl_toolkits/mplot3d/axes3d.py#L1556-L1557

So the input range, assuming -1 to 1, is normalized into a 0 to 1 interval. Here, a positive or negative value would correspond to the same offset from 0.5 (scaled by 2). So a negative offset will give a value less than 0.5 and a positive larger than 0.5. This 0 to 1 interval is then mapped back to a 0.3 to 1 interval and multiplied with the color range in RGB space.

The shade is computed here: https://github.com/matplotlib/matplotlib/blob/cf38c3464ebdbc66eee97222445f9b861ac1a395/lib/mpl_toolkits/mplot3d/axes3d.py#L1550-L1551

where in this case, the default lightsource is https://github.com/matplotlib/matplotlib/blob/cf38c3464ebdbc66eee97222445f9b861ac1a395/lib/mpl_toolkits/mplot3d/axes3d.py#L1547

I cannot really get my head around how to make this simply "invertible" such that the sinusoid plots becomes "symmetric/identical". I believe that another reason may be that the normal is estimated based on three points rather than the complete polygon:
https://github.com/matplotlib/matplotlib/blob/cf38c3464ebdbc66eee97222445f9b861ac1a395/lib/mpl_toolkits/mplot3d/axes3d.py#L1504-L1507

So although the shades of the second plot are the negated version of the first plot, nothing says that there will be corresponding shades (= normals) for the second plot as the normal is approximately computed,

Probably the better solution is to compute the normal more accurately, using all the points of the polygon. It may be hard to do this in general, but as we know that in this case we do have rectangles, it should be somewhat feasible at least. Another option is to split all rectangles into two triangles, but then there are choices regarding which direction to split etc. Yet another option is to split in four triangles, but then we need to determine the best "center point"... So I'd say that probably the first approach is the most feasible here. As long as a rectangle (not being in a plane) returns the same normal with changed sign when swapping the order of the points it should at least be "symmetric" when transposing.

**Comment by oscargus** [comment]
Long answer, but the more I wrote, the more convinced I am that the problem with the sinusoid is that seemingly symmetrically positioned rectangled do have different normals, not just when it comes to the sign, because of only three points being used to determine the normal.

**Comment by oscargus** [comment]
Here is a good figure describing how one can possibly compute the "normal" of a rectangle not in a plane:

https://www.researchgate.net/figure/Normal-definition-of-triangle-and-rectangle-surface-facets_fig1_220677664

**Comment by axil** [comment]
I'm still thinking of Jody's argument about the open-sided cube and slowing coming to a conclusion that this is 'by design': it is something very hard to overcome properly without the actual ray-tracing.

**Comment by oscargus** [comment]
Yes, I also think it is hard to overcome. Especially since index 0, 1, and 3 are indeed the same points it turns out... (I implemented rectangle normals, but it didn't help, and then realized that elements 0, 1, 3 of a sequence `a, b, c, d, a` are indeed `a, b, d` even if it is reversed, but in different order...)

**Comment by oscargus** [comment]
I'll just paste my (useless) rectangle normal code here as I will not commit it to any branch (note that it may be that one should change the sign of one of the cross-product arguments, but that is not the problem anyway):
```
    def _generate_rectangle_normals(self, polygons):
        if isinstance(polygons, np.ndarray):
            assert polygons.shape[-2] == 5
            v1 = polygons[..., 0, :] - polygons[..., 1, :]
            v2 = polygons[..., 1, :] - polygons[..., 2, :]
            v3 = polygons[..., 2, :] - polygons[..., 3, :]
            v4 = polygons[..., 3, :] - polygons[..., 4, :]
        else:
            v1 = np.empty((len(polygons), 3))
            v2 = np.empty((len(polygons), 3))
            v3 = np.empty((len(polygons), 3))
            v4 = np.empty((len(polygons), 3))
            for poly_i, ps in enumerate(polygons):
                v1[poly_i, :] = ps[0, :] - ps[1, :]
                v2[poly_i, :] = ps[1, :] - ps[2, :]
                v3[poly_i, :] = ps[2, :] - ps[3, :]
                v4[poly_i, :] = ps[3, :] - ps[0, :]

        n1 = np.cross(v1, v2)
        n2 = np.cross(v2, v3)
        n3 = np.cross(v3, v4)
        n4 = np.cross(v4, v1)
        s = n1 + n2 + n3 + n4
        return s/np.linalg.norm(s)
```

**Comment by axil** [comment]
Here's an example of a rare person who ran into this problem (without knowing it). He generated the grid in the 'wrong' way and got somewhat subpar results. Left figure has the 'correct' grid, the right one - 'incorrect':
![image](https://user-images.githubusercontent.com/170910/183304666-a98eff96-f702-4e75-b21d-7b8689be966f.png)

https://www.geeksforgeeks.org/3d-surface-plotting-in-python-using-matplotlib/

In this particular case it is not *that* bad, but the 'correct' grid still looks a bit nicer.

**Comment by oscargus** [comment]
I also realize that a rectangle have the same color on both sides, which makes the sign of the normal really important and all my previous approches pointless...

Maybe the solution then is to normalize the rectangles in some way such that it doesn't matter in which order the paths are generated, the same rectangle corners will always give the same path and therefore the same normal and the same shade. Probably one should do this in `plot_surface` to avoid breaking other things where this is potentially used as a feature.

The question is then how to normalize the four corners?

(From an implementation perspective one should be able to enable/disable the normalization to not break earlier plots.)

**Comment by axil** [comment]
@oscargus If you look at the cube from different angles (eg `%matplotlib qt` or `notebook`) you would see that top and bottom faces have different colors.

``` python
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import numpy as np
 
 
# Create axis
axes = [5, 5, 5]
 
# Create Data
data = np.ones(axes, dtype=np.bool)
 
# Control Transparency
alpha = 0.9
 
# Control colour
colors = np.empty(axes + [4], dtype=np.float32)
 
colors[:] = [1, 0, 0, alpha]  # red
 
# Plot figure
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
 
# Voxels is used to customizations of the
# sizes, positions and colors.
ax.voxels(data, facecolors=colors)
```

**Comment by axil** [comment]
@oscargus Maybe it would make sense to detect the 'correctness' of the grid in `plot_surface` and display a warning like:

    To get optimal shading, transpose arguments x, y and z.

in case the grid is 'wrong'?

**Comment by oscargus** [comment]
Yes, that is clearly a good option!

**Comment by oscargus** [comment]
> have different colors.

Yes. I was more referring to "the shaded side" of a surface, not an object. That is, the normal and the shade is computed at draw time, which would basically change the sign of the normal when looking at the "back". Taking the initial example, the two plots could basically be the two different sides of the same plane (which is what happens with the negation, just not at draw time).

**Comment by axil** [comment]
This effect is the consequence of the fact that the shading algorithm is generic, for arbitrary objects, where the same surface can act as either 'top' or 'bottom' face of the object. 

On the other hand `plot_surface` is always the 'top face' as every (x,y) is guaranteed to have a single z. (except if someone is calls it several times to create a complex object).

**Comment by axil** [comment]
Hmm. No, it is not guaranteed (code from [this](https://matplotlib.org/2.0.2/mpl_examples/mplot3d/surface3d_demo2.py) example):
![image](https://user-images.githubusercontent.com/170910/184067958-2c5640a2-e4fc-4334-97f3-b64705393e16.png)
Left: normal pointing outwards, right: normal pointing inwards.
